[patched-9] Port to Qt6, update for C++17/20, and improve compiler/platform support

hjmjohnson opened the pull request:
A summary of all changes from mevislab/master to the current HEAD (slicer-2025-06-13-a2687828):

- Qt6 Porting: Updated code to support Qt6, including replacing deprecated QVariant::Type with QMetaType and adding version checks for Qt5/Qt6 compatibility.
- C++17/20 Modernization: Refactored code to use modern C++ features as required by newer Qt versions (e.g., constexpr, noexcept, alignas, etc.).
- Compiler Detection Updates: Enhanced and updated compiler detection macros in qcompilerdetection.h for better support of recent compilers and platforms.
- Warning and Attribute Macros: Improved handling of compiler warnings and attributes, including support for new C++ attributes like [[nodiscard]], [[maybe_unused]], and [[deprecated]].
- Platform and Feature Checks: Added or updated macros for platform-specific and feature-specific checks, ensuring better cross-platform compatibility.
- General Maintenance: Bug fixes, code cleanup, and improved documentation/comments throughout the codebase.

These changes collectively modernize the codebase, improve compatibility with recent Qt and C++ standards, and enhance maintainability.

jbowler committed:
Qt6: qSort and qStableSort removed (#117)

* replace qSort and qStableSort by 
std::stable_sort and std::sort

jbowler committed:
Qt6: qSort, toList and fromList removal (#118)

This completes the qSort handling in generator/ including related
changes caused by the removal of QSet<>::toList() and QSet<>::fromList()
in Qt6 (obsoleted in Qt5.14).

jbowler committed:
Qt6: QString::SkipEmptyParts moved to Qt (#119)

 - replaced it and added backwards compatibility

jbowler committed:
Qt6: QTextStream compatibility (#120)



- provides compatibility with the removal of the global ::endl
for use with QTextStream; Qt6 requires Qt::endl
- includes Qt6 specific changes necessary to deal with the
replacement of QTextCodec by QStringConverter and the slight changes
to the default setup of QTextStream (to UTF-8 with auto-detect of
UTF-16)

---------

Signed-off-by: John Bowler <jbowler@acm.org>

jbowler committed:
Eliminate warnings about #warning (#122)

The warnings are already only emitted with GCC so changing them to the
GCC diagnostic pragma is safe.

jbowler committed:
Qt6: generator/typesystem.cpp: XML workround (#123)

- works around the need to rewrite the XML parser in
generator/typesystem.cpp by invoking the Qt5 compatibility library;
this is the only thing in the generator which requires it but the
alternative is a complete rewrite of the XML reading code

jbowler committed:
Add constexpr and decltype parsing (#125)

- parses constexpr (in the manner of const) and decltype (in the manner of __typeof) 
  and adds flags for them

jbowler committed:
Handle constexpr, auto (#126)

These changes use the new and existing support inside the parser to
handle constexpr in Qt6 files and to ignore (with a warning) 'auto'
functions.

jbowler committed:
Qt6: fixes for removal of Qt5 types (#127)

Also add the alternate types that are available in Qt5 to replace the
types deprecated in Qt5.14 (allows Qt5.15LTS users to make changes
for compatibility with Qt6.)

jbowler committed:
Fix problems in Qt5.11 generator (#128)

- a previous fix for Qt5.12 breaks pythonqt_generator for 5.11
- includes better checking for precompiled headers and errors out early rather than
  failing when linking the src/ with mysterious missing symbols.

jbowler committed:
Use spaces instead of tabs (#129)

Expand all the tabs in non-binary source files

jbowler committed:
generator: remove anonymous structs (#121)

This is a minimal change to give the three anonymous structs used
in the generator names so that the code conforms to the existing
C++ standards.

jbowler committed:
generator: call Reporthandler::setContext more (#131)

This adds 'context' to ReportHandler warning messages; just a string
saying what is being done.

jbowler committed:
generator: AbstractMetaClassList: correct sorting (#132)

In some places the list was sorted by pointer (resulting in a random
sort) in other cases by name, sometimes with stable_sort, sometimes not.
Now sort by name and maintain do a stable_sort.

mrbean-bremen committed:
Fix handling of right-shift as template bracket

- two consecutive right angle brackets had been handled as right-shift operator
- replace right-shift with two brackets in parser where the context is clear

mrbean-bremen committed:
Ignore variadic template parameters

- ignore default template parameters

jbowler committed:
generator: fix iterator issues with Qt5.14 headers (#142)

const_iterators and other iterators became visible to the generator with
changes in Qt5.14.  This removes the corresponding iterators from the
affected classes.

mrbean-bremen committed:
Make sure the ubuntu artifacts do not overwrite each other

- also fix minor readme error

usiems committed:
add option to print parser errors when they occur

usiems committed:
Keep track of current source code in Debug mode,

also fixed some warnings

usiems committed:
Don't interpret "= default" or "= delete" to mean this is an abstract method

usiems committed:
Avoid currently unsupported use of "noexcept(...)"

usiems committed:
Implement and use Parser::skipFunctionBody

this way we can avoid some parser errors.
This is implemented by looking for matching braces. I hope there is no case where unmatched braces are allowed (strings and comments should be handled by the tokenizer).

usiems committed:
Cover cases where noexcept comes with an expression

usiems committed:
always ignore .vs folder

usiems committed:
Ignore some more macros that lead to parser errors

usiems committed:
Support basic new-style initializers

usiems committed:
Add missing enum used as default value for a method

usiems committed:
Deleted methods need to marked as "invalid" to be handled correctly

usiems committed:
Automatically mark methods with r-value references (&&) as invalid

usiems committed:
Don't create wrappers for cbegin and cend

usiems committed:
Hopefully make generator immune against iterator order changes for typeMap

Previously the includes and generated methods for e.g. QItemSelection would change depending on the order in which the AbstractMetaClass entries were generated, which in turn would change which definition of QList would be seen first.
I now try to mark the "main" definition containing the actual methods and such.
For Qt 5.15 this seems to work at least, but I admit that I still don't understand the generator completely.

usiems committed:
Don't generate wrappers for templated methods in non-template classes

usiems committed:
Simplify skipFunctionBody by only looking at curly braces

YuriUfimtsev committed:
Add sanitizers

YuriUfimtsev committed:
Add a token consistency check between enum and array

usiems committed:
Handle ellipsis in type parser

usiems committed:
Improve debugging code for rewind, also show identifiers

jbowler committed:
Fix: AbstractMetaBuilder::classesTopologicalSorted (#148)

This explicit new/delete, storage leaks (missing delete) and iteration
over a QHash while modifying it (deleting specific keys in an iterator).

usiems committed:
Better handling of "auto" and trailing return types "-> decltype()"

usiems committed:
Better handling of variadic templates

usiems committed:
Skip attribute specifiers like [[maybe_unused]]

YuriUfimtsev committed:
Reduce memory leaks by implementing the AbstractMetaBuilder destructor

usiems committed:
Fix types being rejected because they contain "constexpr"

This leads me to the conclusion that "constexpr" shouldn't be handled like "const". At least for the purpose of wrapper generation I believe "constexpr" can simply be ignored. "volatile" and "mutable" probably as well, at least when it comes to type equality.

usiems committed:
Handle "friend" functions with implementation body...

like the un-scoped functions that they really are

usiems committed:
Treat constexpr as a function modifier, not as a type modifier

usiems committed:
Treat noexcept inside of expressions

(I hope this is the correct place)

usiems committed:
Fix expansion of variadic macros in preprocessor

usiems committed:
Fix handling of elif in preprocessor...

it was totally broken, macros were not expanded

usiems committed:
Improve handling of new-style initializers

usiems committed:
Improve type name handling

usiems committed:
Make parsing more robust against shuffling of qualifiers/specifiers

usiems committed:
Avoid crash if supposed property setter has no arguments

usiems committed:
Avoid crash in lexer by increasing size of token_stream early

The previous code assumed that any parsing step would only add one token at a time, which isn't true. This would lead to crashes if the size doubling threshold was skipped at one point.

usiems committed:
Fix missing includes for namespace items from different files

usiems committed:
Implement methods with lvalue/rvalue references, ...

also moved code into an own method to avoid some code duplication

usiems committed:
Handle (templated) functions with packed parameters

usiems committed:
Support for new-style initializers in expressions

(this doesn't change the generated code at this point, but reduces parsing errors)

usiems committed:
Add some entries to typesystem files missing for Qt6

usiems committed:
Some necessary defines for Qt6

(mostly the same as in parser/rpp/pp-qt-configuration, but for Qt6)

usiems committed:
Add qsizetype and qptrdiff as primitive types

usiems committed:
Add version qualifiers to typesystem XML files

The attributes after-version and before-version restrict the validity of any element to certain Qt versions (after-version means actually including the given version). Additionally there is a new "group" element that only serves the purpose of grouping elements so that the version range doesn't need to repeated. Version strings can be something like "6", "6.5", or "6.5.1".

Some actual elements from the typesystem files have been equipped with these new version attributes for testing purposes.

usiems committed:
Make number of classes per generated file configurable

usiems committed:
Parse inline namespaces,

but treat them like normal namespaces for now

usiems committed:
Parse string literal operator definition,

but it isn't handled specially for now

usiems committed:
Automatically add enums marked with Q_ENUM to TypeDatabase

If this is for a QObject-derived class, they are ignored automatically,
but if it is for a Q_GADGET, they obviously are intended for use with
scripting

usiems committed:
Add new multimedia classes

YuriUfimtsev committed:
Add workflow using generated code (#151)

* generated code is used for PythonQt build
* add CI for the  Qt5. 12 and 5.15

mrbean-bremen committed:
Make sure uploaded wrapper names are different

iakov committed:
Add Dependabot for GH actions

usiems committed:
Don't assume some Qt version if it could not be found/parsed

but allow to specify it on the commandline, just in case

usiems committed:
Fix workflow for MacOS

where the Qt version can't be determined automatically, probably because the include files are inside of frameworks, but that's ok, so we also test giving the version on the command line

usiems committed:
Fix uninitialized member

usiems committed:
Some more defines we don't need to parse

usiems committed:
Q_PROPERTY might contain "REVISION(1, 1)"

usiems committed:
More missing or updated type entries

usiems committed:
Improve parsing of Q_PROPERTY types starting with "const"

usiems committed:
Rename "after-version" attribute to "since-version"

since that better matches the actual behavior

usiems committed:
Also try to build against Qt 6.5

usiems committed:
Handle enum classes

Since the enum values can be duplicates (they only need to be unique in the scope of the enum class) I opted to prefix them with the enum class name (and an underscore) in the generated wrapper.
It would have been nice to use a Python Enum class instead, but after a quick look this seemed to be quite a lot of work.

dependabot[bot] committed:
Bump actions/checkout from 3 to 4 (#158)

* Bump actions/checkout from 3 to 4

Bumps [actions/checkout](https://github.com/actions/checkout) from 3 to 4.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v3...v4)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-type: direct:production
  update-type: version-update:semver-major
...

* Revert action update for oldschool job

- missing libraries on these systems

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: mrbean-bremen <mrbean-bremen@users.noreply.github.com>

usiems committed:
Complete/fix scope for enum class values

usiems committed:
Remove figureOutEnumValues and company

The resulting enumValueRedirection was not used in the PythonQt generator.
Perhaps it was a remnant of the original Java wrapper generator.

usiems committed:
Do not add copy constructor if it was deleted or private

obviously copying the object is explicitly prohibited

usiems committed:
QMutexLocker is a template now, don't know how to handle this

usiems committed:
Fix handling of protected operators

(in this case assignment operators of QEvent-derived classes in Qt6)

usiems committed:
Fix compiler warning (possibly unintended copy)

usiems committed:
Fix QList::swap missing the template argument in the argument type

usiems committed:
I believe it is not necessary to create a shell for QPagedPaintDevice

usiems committed:
Try to fix library names and linking problems

usiems committed:
Make PythonQtTests compile with Qt6

usiems committed:
Do not prepend enum class name for items of enum classes,

rather re-create enum class as a enum class, and later handle enum
classes in PythonQt itself.

usiems committed:
Correct initialization of return value pointer failed

because of changed interface of QVariant in Qt6

usiems committed:
Some types are aliases in Qt6 now

QPair -> std::pair
QVector -> QList
Also there is no space anymore between ">" of template types

usiems committed:
Comparison rules of QVariant have changed in Qt6

usiems committed:
Avoid emitting QStringList* arguments as QStringList<QString>*

usiems committed:
Avoid some compiler warnings

usiems committed:
Skip alignas specifier

mrbean-bremen committed:
Add Windows to CI for checking generated sources

usiems committed:
Enable multi-processor compilation for Visual Studio

usiems committed:
Also enable the XML modules in QtAll project

usiems committed:
Handle QStringView et al as argument/return value

usiems committed:
Filter out methods that do the same (at least in Python)

usiems committed:
Add "enum class" values to the enum type in Python

Enum class values seem to have pretty generic names (as there is no danger
of name clashes in C++, as they must always be prefixed with the enum
class name). So add them to the enum type in Python too, so one can
always write something like QActionGroup.ExclusionPolicy.None instead of
QActionGroup.None.

Ideally we wouldn't add the enum values to the parent class at all, but
I don't want to break compatibility with older versions of PythonQt.

usiems committed:
Escape reserved names by adding "_"

One example I stumbled upon is QActionGroup.ExclusionPolicy.None, which
is a syntax error. So from now on it will be
QActionGroup.ExclusionPolicy.None_.

usiems committed:
Add missing classes

(and in some cases document when classes were introduced)

usiems committed:
webenginewidgets were not compiled yet

mrbean-bremen committed:
Unset INCLUDE env var

- avoids including system headers

mrbean-bremen committed:
Add QtOpenGLWidgets to masterinclude

- add missing modules for qt6 in CI

usiems committed:
Support QByteArrayView arguments + return values

usiems committed:
Suppress some methods of QByteArray and others...

that have become visible again after the argument type has changed from
int to qsizetype (these are the variants that take a const char pointer
and a size argument - this is not supported in Python, instead we use
the variants with QByteArrayView argument)

usiems committed:
Fix misspelled attribute name

mrbean-bremen committed:
Minor documentation updates

mrbean-bremen committed:
Move implementation from main.h into main.cpp

mrbean-bremen committed:
Replace INCLUDE by PYTHONQT_INCLUDE env var

- avoids accidental inclusion of system-defined include dirs

usiems committed:
More typesystem entries (added classes and rejections)

Just so that there are no unknown rejected classes and enums in the log
file for Qt 6.6. The support for some of the new classes might be incomplete.

usiems committed:
Fix irritatingly wrong version number for Qt6

usiems committed:
Remove work-around for package spelling

by putting the correct spelling in the package name in the typesystem
files

mrbean-bremen committed:
Fix potential crash for unknown inner type

- fixes #60

mrbean-bremen committed:
Register container classes with bool

- closes #52

usiems committed:
Create a global namespace wrapper for each package if required

This currently just contains global enums. The content of these wrappers
will be added to the current package, and to the "Qt" package.

Also look in global namespace wrappers for scope-less enum types

YuriUfimtsev committed:
Add type copying when creating a getter to correct memory deallocation

YuriUfimtsev committed:
Change pointers to QShared pointers to decrease memory leaks

YuriUfimtsev committed:
Refactor: change casts and remove toItem() to make it compatible with QSharedPointer

YuriUfimtsev committed:
Add memory deallocation operators

usiems committed:
Add enum values to the enum type object itself in any case...

...not just when it is a enum class in C++. It seems PySide does this too.

usiems committed:
Small documentation updates

- changed default Python version from 2.6 to 3.10

YuriUfimtsev committed:
Remove obsolete things; change type to auto

YuriUfimtsev committed:
Add more shared pointers; remove AbstractMetaType::copy; remove some delete operators and destructors

YuriUfimtsev committed:
Cleanup

dependabot[bot] committed:
Bump actions/upload-artifact from 3 to 4 (#181)

* Bump actions/upload-artifact from 3 to 4

Bumps [actions/upload-artifact](https://github.com/actions/upload-artifact) from 3 to 4.
- [Release notes](https://github.com/actions/upload-artifact/releases)
- [Commits](https://github.com/actions/upload-artifact/compare/v3...v4)

---
updated-dependencies:
- dependency-name: actions/upload-artifact
  dependency-type: direct:production
  update-type: version-update:semver-major
...

* - do not change version for oldschool builds

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: mrbean-bremen <mrbean-bremen@users.noreply.github.com>

dependabot[bot] committed:
Bump actions/setup-python from 4 to 5 (#182)

* Bump actions/setup-python from 4 to 5

Bumps [actions/setup-python](https://github.com/actions/setup-python) from 4 to 5.
- [Release notes](https://github.com/actions/setup-python/releases)
- [Commits](https://github.com/actions/setup-python/compare/v4...v5)

---
updated-dependencies:
- dependency-name: actions/setup-python
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

* do not update setup-python for Windows (won't work with Python 3.6)

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: mrbean-bremen <mrbean-bremen@users.noreply.github.com>

usiems committed:
Add generated bindings for Qt 5.15

usiems committed:
On Ubuntu 22.04 this enum value does not exist (yet)

dependabot[bot] committed:
Bump egor-tensin/cleanup-path from 3 to 4

Bumps [egor-tensin/cleanup-path](https://github.com/egor-tensin/cleanup-path) from 3 to 4.
- [Release notes](https://github.com/egor-tensin/cleanup-path/releases)
- [Commits](https://github.com/egor-tensin/cleanup-path/compare/v3...v4)

---
updated-dependencies:
- dependency-name: egor-tensin/cleanup-path
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump actions/upload-artifact from 3 to 4

Bumps [actions/upload-artifact](https://github.com/actions/upload-artifact) from 3 to 4.
- [Release notes](https://github.com/actions/upload-artifact/releases)
- [Commits](https://github.com/actions/upload-artifact/compare/v3...v4)

---
updated-dependencies:
- dependency-name: actions/upload-artifact
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

YuriUfimtsev committed:
Revert copy method

iakov committed:
Force C++11 for older Qt versions, safe/useless for 5.7+

iakov committed:
Better reference management pattern

iakov committed:
Add PYTHONDEVMODE to improve tests

and other nice options like PYTHONMALLOC=debug_malloc (useful for ASan) and more rigorous warnings handling

usiems committed:
Ignore static_assert in header files

(appeared in QMessageBox in Qt 6.6.2)

usiems committed:
Also register std::pair with the generator

(QPair is more or less deprecated)

usiems committed:
Replace deprecated include qsql.h with qtsqlglobal.h for Qt 6

Fixes #195

nim65s committed:
fix format-security

On Arch, with GCC 13.2.1:

```
make[1]: Entering directory '/home/nim/aur/pythonqt/src/pythonqt-3.5.0/generator'
g++ -c -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -Wp,-D_GLIBCXX_ASSERTIONS -Wno-deprecated-declarations -pedantic -Winit-self -Wuninitialized -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -Wp,-D_GLIBCXX_ASSERTIONS -flto -fno-fat-lto-objects -Wall -Wextra -D_REENTRANT -fPIC -DRXX_ALLOCATOR_INIT_0 -DQT_NO_CAST_TO_ASCII -DQT_NO_DEBUG -DQT_GUI_LIB -DQT_XML_LIB -DQT_CORE_LIB -I. -I. -I. -I/home/nim/aur/pythonqt/src/pythonqt-3.5.0/generator/../common -Iparser -Iparser/rpp -Iparser/rpp -I/usr/include/qt -I/usr/include/qt/QtGui -I/usr/include/qt/QtXml -I/usr/include/qt/QtCore -I. -I/usr/lib/qt/mkspecs/linux-g++ -o main.o main.cpp
main.cpp:108:17: error: format not a string literal and no format arguments [-Werror=format-security]
  108 |         qWarning((reason + "This may cause problems with finding the necessary include files.").toUtf8().constData());
```

iakov committed:
Fix tests
   * `opengl` is not required
   * `CHAR_MIN` is `0` for unsigned `char`. It is implementation specific, actually

iakov committed:
Fix build with PYTHONQT_DEBUG enabled

usiems committed:
6.6.3 introduced a new helper class for QOperatingSystemVersion

nim65s committed:
fix format-security on Q_OS_MAC

nim65s committed:
clean more qWarnings()

nim65s committed:
Update generator/main.cpp

usiems committed:
Allow to set and get the render API in QQuickWindow

mrbean-bremen committed:
Add QVersionNumber to generated code

- needed for QLibraryInfo.version()

iakov committed:
Improve PYTHONQTALL_CONFIG env var handling for CMD (#214)

* Parse PYTHONQTALL_CONFIG value
* Add test for env var PYTHONQTALL_CONFIG
* MSVC2019 for Qt5.15 test

iakov committed:
Remove duplicate entry in build.yml

Fixes build issues https://github.com/MeVisLab/pythonqt/actions/runs/9002631907/job/24731361260#step:11:68

usiems committed:
QPixmapCache::find(QString, QPixmap*) was suppressed

but I don't know why, since the method should simply work (like the
other find variant). And it was reported that the method was previously
(before Qt6) created, even though it should have been suppressed.

iakov committed:
Update install-qt-action to v4

iakov committed:
Update install-qt-action for Win and Mac builds

Migrate to `QT_ROOT_DIR`, remove obsolete `PATH` handling

mrbean-bremen committed:
Add support for Qt 6.7

- use Qt 6.7 for Winodws CI build, 6.6 for ubuntu
  (jurplel/install-qt-action has a problem with that version)
- fixes #215

usiems committed:
Fix a problem with character after comment always added to macro definition

This was especially problematic if the line ended with the comment,
since then the next line would be added, too, even though the current
one didn't end with a '\'

usiems committed:
Fix comma being dropped from macro varargs argument

This prevented QT6_DECL_NEW_OVERLOAD_TAIL from being expanded correctly

usiems committed:
Also move friend function definitions to top-level scope if...

they don't have an implementation.

Often the friend definitions are the only definition there is, and we
translate these definitions (currently always operators) to wrapper
members then.

usiems committed:
Skip attributes that can occur after the enum identifier

(this fixes some parser errors)

usiems committed:
Also translate standalone unary operators (like -, !, or ~)

usiems committed:
Qt 6.7 introduced macros to define comparison operators

that might use the "spaceship" operator for C++20 and onwards.

The generator had trouble to parse these. Simply replace the definitions
with something that works for the generator.

usiems committed:
Qt sometimes uses the Qt::Disambiguated_t parameter (with a default value)

to differentiate new methods with a different return value from the old,
deprecated ones.

usiems committed:
decltype can contain a comma-expression

(this fixes some parser errors)

usiems committed:
Correctly parse string literals with a prefix (e.g. raw string: R"...")

(this fixes some parser errors)

usiems committed:
Ignore Q_GADGET_EXPORT for now, it doesn't help us

(this fixes some parser errors)

usiems committed:
Check that externally defined operators are not added twice

usiems committed:
Handle new classes and enums in Qt 6.7

usiems committed:
Introduce alias names for types

Add an entry "aliases" to a type entry in the type database, these types
will automatically be changed to the main type in the generated code.

This fixes the problem that in Qt 6.7 QAudio enums are used under the
QtAudio alias. These enum names are now automatically mapped to the
QAudio name with the right alias entries in the type database.

This doesn't make QAudio available as QtAudio in Python, though.

usiems committed:
Better diagnostic output

Print the place in the code where the (ambiguous) parser error was
emitted (currently only for GCC and MSVC)

usiems committed:
Fix nullptr crash in error message

usiems committed:
Support for unary + in PythonQt itself

usiems committed:
Check for duplicates at one more place

(should fix compare methods generated twice for Qt 5.12/5.15)

MinyazevR committed:
Change centos archive repository and add node16 support

MinyazevR committed:
Fix memory leak and add memory tests

MinyazevR committed:
Add memory tests with detect_leaks=1 to CI

MinyazevR committed:
Fix issue with handling QT_VERSION in build.yml

MinyazevR committed:
Change PY_VERSION_HEX for use PyConfig

MinyazevR committed:
Fix big with LSan suppresion

MinyazevR committed:
Add ccache to CI

MinyazevR committed:
Fix memory leak with PyDict_Copy

tonka3000 committed:
Add detailed build instructions to Readme (#235)

usiems committed:
Remove use of 'OriginalName' option from code generator...

and instead remove these "rename" entries in the typesystem files, which
were not used because of this option (but triggered
the creation of unneeded wrapper methods in derived classes).

usiems committed:
Filter away static operators in namespaces

This was introduced by Qt 6.7.3, where the static operators + and | on
Qt::Key and Qt::KeyModifiers were moved into the Qt namespace.

usiems committed:
reintroduce the runJavaScript method on QWebEnginePage for Qt 6

albeit without the callback parameter so far

mrbean-bremen committed:
Use latest macOS for CI

- maOS 12 support will soon be removed

Commit:
ENH: add CMake build system support and update CI

add static link support
add CMake build support to PythonQtGenerator
prefer to use ${CMAKE_CURRENT_LIST_DIR}/generated_cpp
add custom targets to automatically run generated commands
Initial support for using auto-generated wrappers
enable testing
add cmake ci
CMake CI Process Validation
Update cmake.yml
adapting more Qt header file include paths
fix _qt_include_prefix

Captured partial cmake rewrite from
```
REMOTE: git@github.com:feihong-gz/pythonqt.git
commit d02ca888ffca2012307bfe9d5e1a176eb95f0cfc
Author: feihong <wangfeihong@pzeda.com>
Date:   2025-02-13 02:08:42 -0500
```

FabioLolix committed:
Fix build with Python 3.13 (#246)

* Fix build with Python 3.13, fixes #245
* Update build.yml to python 3.13

mrbean-bremen committed:
Remove no longer supported CentOS 7 builds

- see https://github.com/actions/runner/issues/2906

dependabot[bot] committed:
Bump hendrikmuhs/ccache-action from 1.2.14 to 1.2.16

Bumps [hendrikmuhs/ccache-action](https://github.com/hendrikmuhs/ccache-action) from 1.2.14 to 1.2.16.
- [Release notes](https://github.com/hendrikmuhs/ccache-action/releases)
- [Commits](https://github.com/hendrikmuhs/ccache-action/compare/v1.2.14...v1.2.16)

---
updated-dependencies:
- dependency-name: hendrikmuhs/ccache-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

usiems committed:
Replace preprocessor with code from simplecpp project

usiems committed:
Set /bigobj option for Visual Studio

Otherwise some object files might get too big when putting all code in
one file (which is useful when comparing with previous results)

usiems committed:
CI: Generate wrappers for Qt 6.8

usiems committed:
Adjustments for new stuff in Qt 6.8

mrbean-bremen committed:
Replace ubuntu 20.04 with ubuntu 24.04

- 20.04 will no longer be supported by GH actions

dependabot[bot] committed:
Bump hendrikmuhs/ccache-action from 1.2.16 to 1.2.17

Bumps [hendrikmuhs/ccache-action](https://github.com/hendrikmuhs/ccache-action) from 1.2.16 to 1.2.17.
- [Release notes](https://github.com/hendrikmuhs/ccache-action/releases)
- [Commits](https://github.com/hendrikmuhs/ccache-action/compare/v1.2.16...v1.2.17)

---
updated-dependencies:
- dependency-name: hendrikmuhs/ccache-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

SchrodingersGat committed:
Update README.md

SchrodingersGat committed:
Update README.md

Remove errant character

usiems committed:
Fix that registered classes are not added to the designated package

if they were implicitly added first (to PythonQt.private), and the
explicit register method isn't given a Python module object, only a name.

Fixes #260

dependabot[bot] committed:
Bump hendrikmuhs/ccache-action from 1.2.17 to 1.2.18

Bumps [hendrikmuhs/ccache-action](https://github.com/hendrikmuhs/ccache-action) from 1.2.17 to 1.2.18.
- [Release notes](https://github.com/hendrikmuhs/ccache-action/releases)
- [Commits](https://github.com/hendrikmuhs/ccache-action/compare/v1.2.17...v1.2.18)

---
updated-dependencies:
- dependency-name: hendrikmuhs/ccache-action
  dependency-version: 1.2.18
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

usiems committed:
Fix conversion of large integers to quint64

Values in the range of 0x7FFFFFFFFFFFFFFF to 0xFFFFFFFFFFFFFFFF caused
an OverflowError when converted to quint64

mrbean-bremen committed:
Fix PythonQt build under Qt 6.9

- add Qt 6.9 to CI

mrbean-bremen committed:
Use Py_REFCNT instead of ob_refcnt in generated code

- ob_refcnt is not guaranteed to be there

hjmjohnson committed:
ENH: Remove QT5 modules from Qt6 QTMODULES

Only search for one of Qt5 or Qt6

Never search for both Qt5 and Qt6 simultaneously.
Requiring user to specify the source bindings to
create.

hjmjohnson committed:
ENH: If not set, choose Release BUILD_TYPE

Prefer to build in Release mode if not set.

hjmjohnson committed:
DOC: Add documentation for installing Qt6

Add initial documentation for how to install
Qt6 on Ubuntu 24.04.

Update README with improved build instructions, CMake usage,
platform-specific notes, and developer hints

hjmjohnson committed:
ENH: Set old QT_VERSION_* variables.

hjmjohnson committed:
PERF:  Don't create temporary QRegularExpression

warning: Don't create temporary QRegularExpression objects. Use a
static QRegularExpression object instead [-Wclazy-use-static-qregularexpression]

Is alerting that the same QRegularExpression object is re-created
every time a function is called or expression is evaluated. This is
inefficient, especially for repeated use, because QRegularExpression has
a non-trivial construction cost (it compiles the regex pattern).

auto match = re.match(content);

hjmjohnson committed:
BUG: Logic bug in contenating strings.

PythonQt/generator/main.cpp:98:60: warning: adding 'bool' to a string does not append to the string [-Wstring-plus-int]
        QString reason = "The QTDIR environment variable " + qtdir.isEmpty() ?
                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
PythonQt/generator/main.cpp:98:60: note: use array indexing to silence this warning
        QString reason = "The QTDIR environment variable " + qtdir.isEmpty() ?
                                                           ^
                         &                                 [                ]
PythonQt/generator/main.cpp:98:78: warning: operator '?:' has lower precedence than '+'; '+' will be evaluated first [-Wparentheses]
        QString reason = "The QTDIR environment variable " + qtdir.isEmpty() ?
                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
PythonQt/generator/main.cpp:98:78: note: place parentheses around the '+' expression to silence this warning
        QString reason = "The QTDIR environment variable " + qtdir.isEmpty() ?
                                                                             ^
                         (                                                  )
PythonQt/generator/main.cpp:98:78: note: place parentheses around the '?:' expression to evaluate it first
        QString reason = "The QTDIR environment variable " + qtdir.isEmpty() ?

hjmjohnson committed:
ENH: Add src to default include directories for PythonQt.h

hjmjohnson committed:
COMP: Remove use of deprecated 'count()' function

PythonQt/generator/parser/name_compiler.cpp:111:46: \
    warning: 'count' is deprecated: Use size() or length() instead.
    [-Wdeprecated-declarations]

hjmjohnson committed:
BUG: Remove deprecated Qt6 paradigms

Replace with backward compatible recommended or
preprocessor conditional for old and new.

Use new paradigms that are backward compatible
when possible.

Use preprocessor conditionals where the syntax
is different between qt5 and qt6.

hjmjohnson committed:
BUG: Fix PyUnicode/PyString usage for Qt6 compatibility

Add preprocessor conditional to use PyUnicode_FromString for Qt6 and PyString_FromString for earlier versions, ensuring compatibility with both Qt5 and Qt6.

hjmjohnson committed:
ENH: Removing obsolete python2 support

`#ifndef PY3K` conditionals and related logic

Removal of PY3k ifdef ifndef not done correctly.

hjmjohnson committed:
REF: Remove outdated `PY_VERSION_HEX` conditional branches

Simplified the code by removing legacy preprocessor conditionals for Python versions < 2.4. Used the consistent `PyMarshal_Write*` API with `Py_MARSHAL_VERSION`.

hjmjohnson committed:
ENH: Define environment variables for Python and Qt versions in CMake

Ensure that the CMake codegen and test environments have the environmental variables set as described in the
Readme.md file.

hjmjohnson committed:
ENH: Update CMake to exclude qtscript_masterinclude.h and improve Qt5/6 compatibility

Explicitly exclude `qtscript_masterinclude.h` from sources. Add compatibility for Qt5 and Qt6 with conditional linking of `Core5Compat` library.

hjmjohnson committed:
ENH: Improve Qt include directory detection in generator CMake

Enhanced logic to handle potential subdirectories like "QtCore" and added fallback for frameworks without "QtCore" suffixed, improving compatibility with various Qt installations.

hjmjohnson committed:
ENH: Add file existence and readability checks for generator inputs

Improve error handling by introducing validation to ensure input files exist and are readable before proceeding, enhancing robustness of the generator.

hjmjohnson committed:
ENH: Add Qt6-specific build_all file and update generator resource handling

Introduce `build_all_qt6.txt` for Qt6 builds and update generator to conditionally copy appropriate resource files. Improved resource copying logic and adjusted CMake commands for enhanced Qt5/6 compatibility.

hjmjohnson committed:
ENH: Simplify Qt version handling in CMake and consolidate component requirements

Removed repetitive Qt version conditionals across multiple CMake files. Consolidated and streamlined required Qt components for both Qt5 and Qt6, ensuring consistency and reducing redundancy.

hjmjohnson committed:
ENH: Add directory existence and readability checks for include paths in generator

Introduce `dirExistsAndReadable` utility to validate include directories before use and enhance debugging output for include resolution in the generator.

hjmjohnson committed:
ENH: Add debug output for object tree dumping in generator

hjmjohnson committed:
ENH: Add compile definitions to enforce safer casting rules in CMake

Introduce compile-time options to prevent dangerous implicit casting in Qt. Add `PythonQt_NO_IMPLICIT_CASTING` for optional stricter casting rules.

hjmjohnson committed:
ENH: Refactor CMakeLists to explicitly define public headers and source files

Replaced `file(GLOB)` usage with explicit lists for source and public header files. Improved clarity and maintainability of project structure in CMake configuration.

hjmjohnson committed:
ENH: Enhance Qt6 include directory resolution

Update generator environment handling

Refactor CMake to improve Qt6 include path detection with duplicate
removal. Add logic for dynamically collecting Qt components and include
directories. Update generator commands to include accessible Qt paths
for more robust environment configuration.

hjmjohnson committed:
ENH: Improve Qt include path resolution macOS

Add specific framework path handling

Refactor include directory parsing to support multiple candidate paths
for headers and introduce macOS-specific logic for resolving
framework-relative header paths. Enhance generator flexibility and
robustness for cross-platform usage.

Push upstream fix via:
https://github.com/danmar/simplecpp/pull/448

hjmjohnson committed:
ENH: Update generated path in CMake to include Qt version-specific suffix

Refine `PYTHONQT_GENERATED_PATH` to append Qt major and minor version, improving compatibility and build separation for different Qt versions.

hjmjohnson committed:
ENH: Refine Qt6 include path resolution for macOS framework support

Enhance include directory parsing to handle macOS-specific framework paths, stripping `.framework/Headers` or `.framework` suffixes when appropriate. Ensure accurate and duplicate-free Qt6 include path collection.

hjmjohnson committed:
BUG: Fix generated path handling in CMake to remove incorrect version-specific suffix

Remove the addition of the Qt version-specific suffix to `PYTHONQT_GENERATED_PATH` as it caused conflicts due to the hardcoded logic in `setupgenerator.cpp`. Add a note for clarification.

hjmjohnson committed:
REF: Replace Q_FOREACH with range-based for loops

The use of Q_FOREACH is deprecated.

Replaced instances of `foreach` with range-based `for`
loops for modernization and code consistency.

hjmjohnson committed:
BUG: Unsupported combination headers caused build error

qprintengine.h:12:1: error: explicit specialization of
   'QMetaTypeId<QMarginsF>' after instantiation

hjmjohnson committed:
REF: Replace `QVariant::Type` with `QMetaType::Type` for Qt compatibility

Update all references from `QVariant::Type` to `QMetaType::Type`
across the generated files for improved alignment with Qt's data
type handling.

hjmjohnson committed:
REF: Replace `qAsConst` with `std::as_const` for C++17 compatibility

hjmjohnson committed:
WIP: Remove old generated files

hjmjohnson committed:
REF: Update verbose/optimize flag handling for Python 3.11 compatibility

Add conditional checks for the `verbose` flag to distinguish
between legacy `Py_VerboseFlag` `Py_OptimizeFlag` and modern `sys.verbose` usage,
ensuring compatibility with Python versions 3.11 and later.

• Python is moving away from global interpreter state.
• Use of global variables like Py_VerboseFlag is not thread-safe.
• Python 3.12+ removes or restricts access to such globals in favor of
  per-interpreter APIs.

hjmjohnson committed:
ENH: Disable deprecated Qt features prior to version 6 in generator

jamesobutler commented:
@hjmjohnson I suspect all the merge conflicts here arise from @jcfr ’s own CMake implementation compared to the other developer that issued a CMake infrastructure PR. As well as @jcfr’s back porting of many commits from the upstream to the CTK patched-9 branch.

jcfr commented:
Thanks @hjmjohnson :pray: , this positioned ourselves well. I will know work on consolidating the changes :rocket:

hjmjohnson committed:
ENH: Add BUILD_EXTENSIONS, BUILD_TESTING, BUILD_EXAMPLES options.

hjmjohnson committed:
COMP: The installed package should be named PythonQt

Rename so that the important installed package is
named PythonQt (instead of Core).

hjmjohnson committed:
ENH: Add configure package options for PythonQt

Configure so that other builds can find PythonQt with find_package

hjmjohnson committed:
ENH: Propogate macro from Python3.12

hjmjohnson committed:
WIP: Comment out items not needed for Slicer

hjmjohnson committed:
ENH: Install all public headers in target.

hjmjohnson committed:
COMP: Need to install all the PythonQt headers.

hjmjohnson committed:
COMP: Add PythonQt_QtBindings.cpp form patch-9 branch.

hjmjohnson committed:
COMP: Backport from patch-9 slicer

 add src/PythonQt_QtBindings.cpp
 add src/PythonQt_QtBindings.h

hjmjohnson committed:
ENH: Add compatibility for Python 3.13 hashing with fallbacks for older versions

Py_HashPointer() is not available in Python 3.12.
It was only added to the public C-API in Python 3.13 (see the “What’s New in 3.13” notes)

Release
Status of a pointer-hash helper
≤ 3.12
Only the private, internal function _Py_HashPointer() exists. It is not exported in the limited- or stable-ABI headers and may disappear or change without notice.
3.13+
Py_HashPointer(const void *ptr) becomes a public API entry in <Python.h>.

hjmjohnson committed:
COMP: Remove _Py_HashPointer as undefined symbol.

hjmjohnson commented:
@jcfr, Please let me know when this branch can be deleted.

hjmjohnson commented:
@jcfr Abandoning this PR. PythonQt was updated from other efforts.
